projects/update_ping_ready_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,347 @@
+    Target channel mismatch: expected nightly-cck-rambler got nightly:	0.002%
+
+
+The vast majority of the data in the payload seems reasonable (99.71%). However, a handful of `update` pings are reporting a `targetBuildId` which is older than the current build reported by the ping's environment. Some pings are also reporting a `targetChannel` different than the one in the environment. These issues do not represent a problem, as most of the data is correct.
+
+## Check that we receive one ping per client and target update
+For each ping, build a key with the client id and the target update details. Since we expect to have exactly one ping for each update bundle marked as *ready*, we don't expect duplicate keys.
+
+
+```python
+update_dupes = deduped_subset.map(lambda p: ((p.get("clientId"),
+                                              p.get("payload/targetChannel"),
+                                              p.get("payload/targetVersion"),
+                                              p.get("payload/targetBuildId")), 1)).countByKey()
+
+print("Percentage of pings related to the same update (for the same client):\t{:.3f}%"\
+      .format(pct(sum([v for v in update_dupes.values() if v > 1]), deduped_count)))
+```
+    Percentage of pings related to the same update (for the same client):	1.742%
+
+
+Apparently we're receiving `update` pings with different `documentId` related to the same target update bundle, for a few clients. The 1.74% seems low enough to not be a problem at this stage.
+
+## Validate the submission delay
+
+
+```python
+delays = deduped_subset.map(lambda p: calculate_submission_delay(p))
+```
+
+```python
+setup_plot("'update' ('ready') ping submission delay CDF",
+           MAX_DELAY_S / HOUR_IN_S, area_border_x=1.0)
+
+plot_cdf(delays\
+         .map(lambda d: d / HOUR_IN_S if d < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+         .collect(), label="CDF", linestyle="solid")
+
+plt.show()
+```
+
+
+![png](images/output_21_0.png)
+
+
+Almost all of the `update` ping are submitted within an hour from the update being ready.
+
+## Make sure that the volume of incoming `update` pings is reasonable
+This is a tricky one. The `update` ping with `reason = "ready"` is sent as soon as an update package is downloaded, verified and deemed ready to be applied. However, nothing guarantees that the update is immediately (or ever) applied. To check if the volume of `update` pings is in the ballpark, we can:
+
+1. Get a list of client ids for a specific target update build id '20170809xxxxxx'.
+2. Get the `main-ping` for that version of Firefox.
+3. Check how many clients from the list at (1) are in the list at (2).
+
+**Step 1** - Get the list of client ids updating to build '20170809xxxxxx'
+
+
+```python
+TARGET_BUILDID_MIN = '20170809000000'
+TARGET_BUILDID_MAX = '20170809999999'
+
+update_candidates =\
+    deduped_subset.filter(lambda p: TARGET_BUILDID_MIN <= p.get("payload/targetBuildId") <= TARGET_BUILDID_MAX)
```

Comment:
For clarity on the edges it might be better to have the TARGET_BUILDID_{MIN,MAX} be the same as the parameters you pass into the `.where(appBuildId...)` clause